File: 2026-03/src/day_08/boj_20304/Main.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    int n, m;
    cin >> n >> m;

    queue<int> q;
    vector<bool> vis(1 + n);

    while (m--) {
        int x;
        cin >> x;

        q.push(x);
        vis[x] = true;
    }

    int dist = 0;

    while (!q.empty()) {
        int sz = q.size();

        while (sz--) {
            int cur = q.front();
            q.pop();

            for (int i = 0; i < 20; i++) {
                int nxt = cur ^ (1 << i);
                if (nxt > n || vis[nxt]) continue;

                q.push(nxt);
                vis[nxt] = true;
            }
        }

        dist++;
    }

    cout << dist - 1;
}

File: 2026-03/src/day_08/boj_20304/Main.java
package day_08.boj_20304;

import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int n = Integer.parseInt(br.readLine());
        int m = Integer.parseInt(br.readLine());

        Queue<Integer> q = new ArrayDeque<>();
        boolean[] vis = new boolean[1 + n];

        st = new StringTokenizer(br.readLine());
        while (m-- > 0) {
            int x = Integer.parseInt(st.nextToken());
            q.offer(x);
            vis[x] = true;
        }

        int dist = 0;

        while (!q.isEmpty()) {
            int size = q.size();

            while (size-- > 0) {
                int cur = q.poll();

                for (int i = 0; i < 20; i++) {
                    int nxt = cur ^ (1 << i);
                    if (nxt > n || vis[nxt]) continue;

                    q.offer(nxt);
                    vis[nxt] = true;
                }
            }

            dist++;
        }

        System.out.println(dist - 1);
    }
}
